tools/lib/src/device.dart

```diff
@@ -0,0 +1,381 @@
+      return 1;
+    });
+
+    await for (final String line in streamLines) {
+      lastLine = line;
+      print(line);
+    }
+
+    final List<String> errors = <String>[];
+    if (timedOut) {
+      errors.add('Timeout expired. The test may need more time to finish. '
+          'If you expect the test to finish before timeout, check if the tests '
+          'require device screen to be awake or if they require manually '
+          'clicking the UI button for permissions.');
+    } else if (lastLine.startsWith('No tests ran')) {
+      errors.add(
+          'Missing integration tests (use --exclude if this is intentional).');
+    } else if (lastLine.startsWith('No devices found')) {
+      errors.add('Device was disconnected during test.');
+    } else {
+      final RegExpMatch? match = _logPattern.firstMatch(lastLine);
+      if (match == null || match.group(2) == null) {
+        throw Exception('Log message is not parsed correctly.');
+      } else if (!match.group(2)!.startsWith('All tests passed!')) {
+        errors.add('flutter-tizen test integration_test failed, see the output '
+            'above for details.');
+      }
+    }
+
+    return errors.isEmpty
+        ? PackageResult.success()
+        : PackageResult.fail(errors);
+  }
+}
+
+/// A reference to a Tizen emulator device.
+class EmulatorDevice extends Device {
+  /// Creates a new reference to a Tizen emulator device.
+  ///
+  /// Valid characters for a device name are [A-Za-z0-9-_].
+  factory EmulatorDevice(
+    String name,
+    Profile profile, {
+    required TizenSdk tizenSdk,
+    ProcessRunner processRunner = const ProcessRunner(),
+  }) {
+    final EmulatorDevice emulatorDevice = EmulatorDevice._(
+      name,
+      profile,
+      tizenSdk: tizenSdk,
+      processRunner: processRunner,
+    );
+    emulatorDevice._id = emulatorDevice._findId();
+    emulatorDevice._pid = findEmulatorPid(name);
+    return emulatorDevice;
+  }
+
+  EmulatorDevice._(
+    String name,
+    Profile profile, {
+    required TizenSdk tizenSdk,
+    ProcessRunner processRunner = const ProcessRunner(),
+  }) : super._(
+          name,
+          profile,
+          tizenSdk: tizenSdk,
+          processRunner: processRunner,
+        );
+
+  String? _pid;
+
+  @override
+  bool get isConnected {
+    _id = _findId();
+    _pid = findEmulatorPid(name);
+    return _id != null && _pid != null;
+  }
+
+  @override
+  bool get isEmulator => true;
+
+  /// Checks whether the emulator with [name] and [profile] exists
+  /// in the Emulator Manager.
+  bool get exists {
+    final io.ProcessResult result =
+        _processRunner.runSync(_tizenSdk.emCli.path, <String>['list-vm']);
+    if (result.exitCode != 0) {
+      throw ToolExit(result.exitCode);
+    }
+
+    final List<String> emulatorNames =
+        LineSplitter.split((result.stdout as String).trim())
+            .map((String name) => name.trim())
+            .toList();
+    return emulatorNames.contains(name);
```

Comment:
Is the purpose of the first `trim()` to ignore any leading and trailing new line? Otherwise you can remove it as spaces are already being removed by the next line (`name.trim()`).